Remove Context dependency from Tensor class

dzhulgakov opened the pull request:
Summary:
Removes reference to Context proper and instead adds a bool argument for async copy (the same as `copy_`)

For CopyFrom - I haven't tweaked all callsites yet. Instead I rely on a terrible hack that pointer to context is implicitly converted to bool when passed, haha :) It's not a good code and I propose to fix it in a follow up diff (maybe using clangr tooling).

Differential Revision: D13117981

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -904,9 +904,9 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+   * 'async' parameter triggers async copy for CUDA tensors
    */
-  void CopyFrom(const TensorImpl& src, at::BaseContext* context = nullptr) {
+  void CopyFrom(const TensorImpl& src, bool async = false) {
```
Nit: In PyTorch, we call this `non_blocking`, because `async` is a reserved keyword in Python so we can't use it as a keyword argument. Just a thing to be aware of :)

ezyang commented on caffe2/core/context_gpu.cu:
```diff
@@ -72,6 +72,74 @@ REGISTER_CONTEXT(DeviceType::CUDA, caffe2::CUDAContext);
+    void* dst,
+    Device dst_device) {
+  // FIXME: verify that the CUDA handles copy from device to device correctly
+  // even without SetDevice()
```
Well, at least this simple test program doesn't error:

```
#include <stdio.h>
#include <cuda.h>
#include <cuda_runtime_api.h>

#include <iostream>

int main()
{
  void* src;
  void* dst;
  cudaSetDevice(1);
  std::cerr << "malloc src " << cudaMalloc(&src, 100) << "\n";
  cudaSetDevice(2);
  std::cerr << "malloc dst " << cudaMalloc(&dst, 100) << "\n";
  cudaSetDevice(0);
  std::cerr << "memcpyAsync " << cudaMemcpyAsync(dst, src, 100, cudaMemcpyDeviceToDevice) << "\n";
  std::cerr << "sync " << cudaDeviceSynchronize() << "\n";
  return 0;
}
```

ezyang commented on caffe2/core/context_gpu.cu:
```diff
@@ -72,6 +72,74 @@ REGISTER_CONTEXT(DeviceType::CUDA, caffe2::CUDAContext);
+    Device dst_device) {
+  // FIXME: verify that the CUDA handles copy from device to device correctly
+  // even without SetDevice()
+  // TODO: verify whether source or dest device should be a priority
```
Not sure what this means. Priority stream?

mruberry replied:
Not me. Maybe @ngimel knows.

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -937,48 +937,32 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+        // Specifically, we might need to switch to a different context device
+        // here explicitly to avoid relying on user synchronizing things
+        // properly.
+        at::CopyBytes(
```
Can we avoid putting this function name in the top-level namespace?

ezyang commented on aten/src/ATen/core/context_base.cpp:
```diff
@@ -9,6 +11,47 @@ C10_DEFINE_TYPED_REGISTRY(
     std::unique_ptr,
     at::Device);
 
+static CopyBytesFunction g_copy_bytes[2][COMPILE_TIME_MAX_DEVICE_TYPES]
```
The doc here should say what 0 means and what 1 means.

ezyang reviewed (approved):
okey dokey

Commit:
double dispatch for copybytes

Differential Revision: D13117987

fbshipit-source-id: 4692cb38924e062727c5b3ef41841b8cea2c1439

Commit:
Remove Context dependency from Tensor class (#14269)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/14269

Removes reference to Context proper and instead adds a bool argument for async copy (the same as `copy_`)

For CopyFrom - I haven't tweaked all callsites yet. Instead I rely on a terrible hack that pointer to context is implicitly converted to bool when passed, haha :) It's not a good code and I propose to fix it in a follow up diff (maybe using clangr tooling).

Reviewed By: ezyang

Differential Revision: D13117981

fbshipit-source-id: 46227a56e41c691a602f11c6701237dd56f1c938
